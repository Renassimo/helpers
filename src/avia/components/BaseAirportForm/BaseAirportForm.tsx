import { useState, KeyboardEvent, useEffect, ReactNode } from 'react';

import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';
import LabelIcon from '@mui/icons-material/Label';
import TitleIcon from '@mui/icons-material/Title';
import MyLocationIcon from '@mui/icons-material/MyLocation';

import useStateValue from '@/common/hooks/useStateValue';

import AirportCard from '@/avia/components/AirportCard';

import AviaInput from '@/avia/components/AviaInput';

import useAlerts from '@/common/hooks/alerts/useAlerts';
import geoLocation from '@/common/utils/geoLocation';

import { Avia } from '@/avia/types/avia';

const BaseAirportForm = ({
  title = 'Airport',
  airportsResult,
  adornments,
}: {
  title: string;
  airportsResult: Avia.AirportsResult;
  adornments?: ReactNode;
}) => {
  const { createErrorAlert } = useAlerts();

  const {
    airports,
    chosenAirport,
    retreiveAirports,
    chooseAirport,
    clearChosenAirport,
    loading,
  } = airportsResult;

  const [airportValue, setAirportValue] = useStateValue<string>('');
  const [by, setBy] = useState<'code' | 'text' | 'location'>('code');
  const [locationLoading, setLocationLoading] = useState(false);

  useEffect(() => {
    if (airportValue.length > 4 && by === 'code') setBy('text');
  }, [airportValue]);

  const searchByLocation = async () => {
    setLocationLoading(true);
    geoLocation(
      async (args) =>
        await retreiveAirports({
          lat: String(args.coords.latitude),
          lon: String(args.coords.longitude),
        }),
      ({ message }) => createErrorAlert(message),
      createErrorAlert,
      () => {
        setAirportValue('');
        setBy('location');
      }
    );
    setLocationLoading(false);
  };

  const searchByCode = async () => {
    setBy('code');
    await retreiveAirports({
      code: airportValue,
    });
  };

  const searchByText = async () => {
    setBy('text');
    await retreiveAirports({
      text: airportValue,
    });
  };

  const onEnter = (event: KeyboardEvent) => {
    if (event.key === 'Enter') {
      if (airportValue.length > 4) {
        searchByText();
      } else if (airportValue.length > 2) {
        searchByCode();
      }
    }
  };

  const inputDisabled = loading || locationLoading;
  const textingDisabled = inputDisabled || airportValue.length < 3;

  return (
    <Box mb={1}>
      {!chosenAirport && (
        <AviaInput
          title={title}
          placeholder={`${title} IATA or ICAO code`}
          label={title}
          value={airportValue}
          setValue={setAirportValue}
          disabled={inputDisabled}
          onKeyDown={onEnter}
          adornments={
            <>
              <IconButton
                size="small"
                color={by === 'code' ? 'primary' : undefined}
                onClick={searchByCode}
                disabled={textingDisabled}
                aria-label="By Code"
              >
                <LabelIcon />
              </IconButton>
              <IconButton
                size="small"
                color={by === 'text' ? 'primary' : undefined}
                onClick={searchByText}
                disabled={textingDisabled}
                aria-label="By Text"
              >
                <TitleIcon />
              </IconButton>
              <IconButton
                size="small"
                color={by === 'location' ? 'primary' : undefined}
                onClick={searchByLocation}
                disabled={inputDisabled}
                aria-label="By Location"
              >
                <MyLocationIcon />
              </IconButton>
              {adornments}
            </>
          }
        />
      )}
      {airports && !chosenAirport && (
        <Box display="flex" overflow="auto">
          {airports.map((airport, index) => (
            <Box
              key={airport.id + index}
              mr={1}
              ml={index === 0 ? 1 : undefined}
            >
              <AirportCard
                data={airport}
                onClick={() => chooseAirport(airport.id)}
              />
            </Box>
          ))}
        </Box>
      )}
      {chosenAirport && (
        <AirportCard
          data={chosenAirport}
          onClick={clearChosenAirport}
          title={title}
          chosen
        />
      )}
    </Box>
  );
};

export default BaseAirportForm;
