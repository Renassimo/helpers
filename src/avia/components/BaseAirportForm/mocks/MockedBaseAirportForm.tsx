const MockedBaseAirportForm = jest.fn(({ title }) => (
  <div>MockedBaseAirportForm - {title}</div>
));

export default MockedBaseAirportForm;
