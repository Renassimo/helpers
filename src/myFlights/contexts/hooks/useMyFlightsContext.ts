import { useContext } from 'react';

import MyFlightsContext from '@/myFlights/contexts/MyFlightsContext';

const useMyFlightsContext = () => useContext(MyFlightsContext);

export default useMyFlightsContext;
