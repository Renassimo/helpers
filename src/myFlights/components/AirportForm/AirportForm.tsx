import { useState, KeyboardEvent, useEffect } from 'react';

import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';
import LabelIcon from '@mui/icons-material/Label';
import TitleIcon from '@mui/icons-material/Title';
import MyLocationIcon from '@mui/icons-material/MyLocation';

import useInputValue from '@/common/hooks/useInputValue';

import useMyFlightsContext from '@/myFlights/contexts/hooks/useMyFlightsContext';

import AirportCard from '@/myFlights/components/AirportCard';

import AviaInput from '@/avia/components/AviaInput';

import useAlerts from '@/common/hooks/alerts/useAlerts';
import geoLocation from '@/common/utils/geoLocation';

const AirportForm = ({ title }: { title: 'Origin' | 'Destination' }) => {
  const { createErrorAlert } = useAlerts();

  const key = title === 'Origin' ? 'originsResult' : 'destinationsResult';
  const {
    [key]: {
      airports,
      chosenAirport,
      retreiveAirports,
      chooseAirport,
      clearChosenAirport,
      loading,
    },
  } = useMyFlightsContext();

  const [airportValue, setAirportValue] = useInputValue<string>('');
  const [by, setBy] = useState<'code' | 'text' | 'location'>('code');
  const [locationLoading, setLocationLoading] = useState(false);

  useEffect(() => {
    if (airportValue.length > 4 && by === 'code') setBy('text');
  }, [airportValue]);

  const searchByLocation = async () => {
    setLocationLoading(true);
    geoLocation(
      async (args) =>
        await retreiveAirports({
          lat: String(args.coords.latitude),
          lon: String(args.coords.longitude),
        }),
      ({ message }) => createErrorAlert(message),
      createErrorAlert,
      () => {
        setAirportValue('');
        setBy('location');
      }
    );
    setLocationLoading(false);
  };

  const searchByCode = async () => {
    setBy('code');
    await retreiveAirports({
      code: airportValue,
    });
  };

  const searchByText = async () => {
    setBy('text');
    await retreiveAirports({
      text: airportValue,
    });
  };

  const onEnter = (event: KeyboardEvent) => {
    if (event.key === 'Enter') {
      if (airportValue.length > 4) {
        searchByText();
      } else if (airportValue.length > 2) {
        searchByCode();
      }
    }
  };

  const inputDisabled = loading || locationLoading;
  const textingDisabled = inputDisabled || airportValue.length < 3;

  return (
    <Box mb={1}>
      {!chosenAirport && (
        <AviaInput
          title={title}
          placeholder={`${title} IATA or ICAO code`}
          label={title}
          value={airportValue}
          setValue={setAirportValue}
          disabled={inputDisabled}
          onKeyDown={onEnter}
          adornments={
            <>
              <IconButton
                size="small"
                color={by === 'code' ? 'primary' : undefined}
                onClick={searchByCode}
                disabled={textingDisabled}
                aria-label="By Code"
              >
                <LabelIcon />
              </IconButton>
              <IconButton
                size="small"
                color={by === 'text' ? 'primary' : undefined}
                onClick={searchByText}
                disabled={textingDisabled}
                aria-label="By Text"
              >
                <TitleIcon />
              </IconButton>
              <IconButton
                size="small"
                color={by === 'location' ? 'primary' : undefined}
                onClick={searchByLocation}
                disabled={inputDisabled}
                aria-label="By Location"
              >
                <MyLocationIcon />
              </IconButton>
            </>
          }
        />
      )}
      {airports && !chosenAirport && (
        <Box display="flex" overflow="auto">
          {airports.map((airport, index) => (
            <Box
              key={airport.id + index}
              mr={1}
              ml={index === 0 ? 1 : undefined}
            >
              <AirportCard
                data={airport}
                onClick={() => chooseAirport(airport.id)}
              />
            </Box>
          ))}
        </Box>
      )}
      {chosenAirport && (
        <AirportCard
          data={chosenAirport}
          onClick={clearChosenAirport}
          title={title}
          chosen
        />
      )}
    </Box>
  );
};

export default AirportForm;
