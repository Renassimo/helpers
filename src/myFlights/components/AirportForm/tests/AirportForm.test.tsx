import renderWithTheme from '@/common/tests/helpers/renderWithTheme';

import useMyFlightsContext from '@/myFlights/contexts/hooks/useMyFlightsContext';

import BaseAirportForm from '@/avia/components/BaseAirportForm';

import MockedBaseAirportForm from '@/avia/components/BaseAirportForm/mocks';

import AirportForm from '../AirportForm';

jest.mock('@/avia/components/BaseAirportForm');
jest.mock('@/myFlights/contexts/hooks/useMyFlightsContext');

describe('AirportForm', () => {
  beforeEach(() => {
    (useMyFlightsContext as unknown as jest.Mock).mockImplementation(
      jest.fn(() => ({
        originsResult: 'originsResult',
        destinationsResult: 'destinationsResult',
      }))
    );
    (BaseAirportForm as unknown as jest.Mock).mockImplementation(
      MockedBaseAirportForm
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders successfully', () => {
    // Arange
    // Act
    const { baseElement } = renderWithTheme(<AirportForm title="Origin" />);
    // Assert
    expect(baseElement).toMatchSnapshot();
    expect(BaseAirportForm).toBeCalledWith(
      { airportsResult: 'originsResult', title: 'Origin' },
      {}
    );
  });

  describe('when titile is Destination', () => {
    test('renders successfully', () => {
      // Arange
      // Act
      const { baseElement } = renderWithTheme(
        <AirportForm title="Destination" />
      );
      // Assert
      expect(baseElement).toMatchSnapshot();
      expect(BaseAirportForm).toBeCalledWith(
        { airportsResult: 'destinationsResult', title: 'Destination' },
        {}
      );
    });
  });
});
