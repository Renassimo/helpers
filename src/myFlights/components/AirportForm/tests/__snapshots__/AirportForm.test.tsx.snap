// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirportForm renders successfully 1`] = `
<body>
  <div>
    <div>
      MockedBaseAirportForm - 
      Origin
    </div>
  </div>
</body>
`;

exports[`AirportForm when titile is Destination renders successfully 1`] = `
<body>
  <div>
    <div>
      MockedBaseAirportForm - 
      Destination
    </div>
  </div>
</body>
`;
